### PR TITLE
[Table] Better column reordering

### LINKED ISSUE #1189: FR Blueprint Table: Add drag handle into header cell
Currently you can't just drag a column. You have to select it, release the mouse, and then re-drag. Our users are finding this confusing.

Would it be possible to add a prop that changes the behavior of the table, so that you can drag columns without selecting them first?

Maybe if the selection mode is set to disable selection of columns, then the table could behave like this automatically.

### LINKED ISSUE #1242: [Table] Add reorder handle to column header cells when reordering enabled
**Problem**
When tasked with reordering a single column in the `Table`, users don't realize that you have to click-select the column first, _then_ reorder it via click-dragging.

**Proposal**
Add drag-reorder handles to `ColumnHeaderCell`s that, when click-dragged, always triggering reordering regardless of the column's selected state.

### LINKED ISSUE #1273: [Table] Fix behavior when reordering multiple disjoint selections
### Bug report

- __Package version(s)__: v1.21.0
- __Browser and OS versions__: Chrome 59 on Mac OS X Sierra

#### Steps to reproduce

1. `Cmd + Click` to select multiple disjoint columns.
1. Drag to reorder them.

#### Actual behavior

Not all of the columns come along for the reordering ride, and the result is a weird selected state after the fact (particularly if you reorder a column that was _not_ the most recently selected one).

_Example:_ If you reorder a column that _was_ the most recently selected one:
reorder-disjoint2

_Example:_ If you reorder a columns that was _not_ the most recently selected one:
reorder-disjoint

#### Expected behavior

(Suggestion:) all disjoint selections are joined into one, then the clump is reordered to the new index?

### LINKED ISSUE #1284: [Table] Reordering appears possible without selection
A table can have reordering turned on, but selection turned off. This leads to users being able to get the grab cursor, and appear to drag the column, but then when they release, nothing changes. 

reorder

P3 because apps should know not to have selection without reordering, and docs exist noting this
Would be nice to bake this in more though

### PR DESCRIPTION
**Pull request by cmslewis**
#### Checklist

- [x] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:

Improvements to column reordering:
- Reorderable columns now _always_ have a reorder handle, even if the interaction bar is disabled.
- Columns can now be reordered even when selection is disabled.
- Reordering disjoint selections is now much more intuitive.

#### Reviewers should focus on:

Do you like the visual design of the reorder handle? And the interactions?

#### Screenshot

Reordering a column with selection disabled:

Reordering a selected column:

Reordering a multi-column selection:

And of course, all of this works when the interaction bar is visible too:

### CODE REVIEW

**Comment by adidahiya** [comment]
Interactions feel great!

> Reordering disjoint selections is now much more intuitive.

I didn't notice how unintutive it was earlier. It is certainly better now. Nice work 👍 

Since the table now only moves one of the columns in a disjoint selection and disregards the others, is it possible to make it so that only that one columns (the one where you started the drag) is highlighted after you start a drag? At this point:

image

If that's too much trouble then don't worry about it for now.

File: packages/table/src/common/classes.ts [hunk 1]
```diff
@@ -17,10 +17,12 @@ export const TABLE_CELL_LEDGER_EVEN = "bp-table-cell-ledger-even";
 export const TABLE_CELL_LEDGER_ODD = "bp-table-cell-ledger-odd";
 export const TABLE_COLUMN_HEADER_TR = "bp-table-column-header-tr";
 export const TABLE_COLUMN_HEADERS = "bp-table-column-headers";
+export const TABLE_COLUMN_HEADER_CELL = "bp-table-column-header-cell";
+export const TABLE_COLUMN_HEADER_CELL_HAS_REORDER_HANDLE = "bp-table-column-header-cell-has-reorder-handle";
```
Comment (adidahiya) [inline packages/table/src/common/classes.ts]: yikes this is pretty long. I would be cool with simply `bp-table-has-reorder-handle` and `bp-table-has-interaction-bar`

File: packages/table/src/interactions/_interactions.scss [hunk 2]
```diff
@@ -48,24 +52,37 @@ $resize-handle-dragging-color: $pt-intent-primary !default;
...
 
+// we either need a long selector or a deeply nested selector here. prefer the former.
+// stylelint-disable max-line-length
+.bp-table-column-header-cell.bp-table-column-header-cell-has-reorder-handle:not(.bp-table-column-header-cell-has-interaction-bar) {
+  .bp-table-column-name-text {
+    padding-left: $reorder-handle-width;
+  }
+
+  // keep the editable cell input flush with the cell's left border
+  .bp-table-column-name .bp-table-editable-name.pt-editable-text::before {
```
Comment (giladgray) [inline packages/table/src/interactions/_interactions.scss]: _this_ is both long _and_ deeply nested 😨 
must it be this way?

File: packages/table/test/columnHeaderCellTests.tsx [hunk 3]
```diff
@@ -143,22 +143,18 @@ describe("<ColumnHeaderCell>", () => {
     });
 
     describe("Reorder handle", () => {
-        const REORDER_HANDLE_CLASS = "reorder-handle";
+        const REORDER_HANDLE_CLASS = "bp-table-reorder-handle-target";
```
Comment (giladgray) [inline packages/table/test/columnHeaderCellTests.tsx]: prefer not using `bp-table-` prefix for test-only class. be obvious that this is not a real `Table` class, just a hook used for these tests. i was wondering if it belongs in `Classes`, then I found its usage in `mount()` below.

File: packages/table/test/tableTests.tsx [hunk 4]
```diff
@@ -689,150 +689,112 @@ describe("<Table>", () => {
...
-        const OFFSET_X = (NEW_INDEX + LENGTH) * WIDTH_IN_PX;
+        const OFFSET_X = (NEW_INDEX + LENGTH) * WIDTH_IN_PX - REORDER_HANDLE_WIDTH_IN_PX;
         const OFFSET_Y = (NEW_INDEX + LENGTH) * HEIGHT_IN_PX;
 
-        let onColumnsReordered: Sinon.SinonSpy;
-        let onRowsReordered: Sinon.SinonSpy;
-        let onSelection: Sinon.SinonSpy;
+        const onColumnsReordered: Sinon.SinonSpy = sinon.spy();
+        const onRowsReordered: Sinon.SinonSpy = sinon.spy();
+        const onSelection: Sinon.SinonSpy = sinon.spy();
```
Comment (giladgray) [inline packages/table/test/tableTests.tsx]: prob don't need `: Sinon.SinonSpy` anymore

File: packages/table/test/tableTests.tsx [hunk 5]
```diff
@@ -689,150 +689,112 @@ describe("<Table>", () => {
...
         });
 
         it("Doesn't work on rows if there is no selected region defined yet", () => {
             const table = mountTable({
                 isColumnReorderable: true,
                 onColumnsReordered,
             });
-            getTableHeader(getColumnHeadersWrapper(table), 0)
+            const headerCell = getHeaderCell(getColumnHeadersWrapper(table), 0);
+            headerCell
```
Comment (giladgray) [inline packages/table/test/tableTests.tsx]: revert? this variable doesn't help

File: packages/table/test/columnHeaderCellTests.tsx (hunk 3 above)
Comment (cmslewis) [inline packages/table/test/columnHeaderCellTests.tsx]: It's not a test-only class.

...I should use `Classes.TABLE_REORDER_HANDLE_TARGET`.

File: packages/table/test/tableTests.tsx (hunk 5 above)
Comment (cmslewis) [inline packages/table/test/tableTests.tsx]: Sure

**Comment by cmslewis** [comment]
Read through the docs changes with @tgreenwatts. Ready to merge this as soon as the build finishes. 👍

**Pull request merged**